lib/scholar/linear/logistic_regression.ex

```diff
@@ -0,0 +1,192 @@
+    end
+
+    if is_number(opts[:lr]) and opts[:lr] < 0 do
+      raise ArgumentError, "Learning rate must be a positive number"
+    end
+
+    if is_integer(opts[:iterations]) and opts[:iterations] <= 0 do
+      raise ArgumentError, "Number of iterations must be a positive integer"
+    end
+
+    if opts[:num_classes] < 3, do: fit_binary(x, y, opts), else: fit_multinomial(x, y, opts)
+  end
+
+  # Binary logistic regression
+
+  defnp fit_binary(x, y, opts \\ []) do
+    opts = keyword!(opts, iterations: 1000, lr: 0.01, num_classes: 2, one_hot: false)
+
+    iterations = opts[:iterations]
+    lr = opts[:lr]
+    y = if opts[:one_hot], do: Nx.argmax(y, axis: 1), else: y
+
+    {_m, n} = x.shape
+    coeff = Nx.broadcast(Nx.tensor(0, type: {:f, 32}), {n})
+
+    {_, _, _, _, _, final_coeff, final_bias} =
+      while {iter = 0, x, y, lr, iterations, coeff, bias = Nx.tensor(0, type: {:f, 32})},
+            Nx.less(iter, iterations) do
+        {coeff, bias} = update_coefficients(x, y, {coeff, bias}, lr)
+        {iter + 1, x, y, lr, iterations, coeff, bias}
+      end
+
+    %LogisticRegression{coefficients: final_coeff, bias: final_bias, mode: 0}
+  end
+
+  # Function computes one-hot encoding
+
+  defnp one_hot_encoding(labels, num_classes) do
+    Nx.equal(Nx.new_axis(labels, -1), Nx.iota({1, num_classes}))
+  end
+
+  # Multinomial logistic regression
+
+  defnp fit_multinomial(x, y, opts) do
+    {_m, n} = x.shape
+
+    opts =
+      keyword!(opts,
+        iterations: 1000,
+        lr: 0.01,
+        num_classes: 1,
+        one_hot: false
+      )
+
+    iterations = opts[:iterations]
+    lr = opts[:lr]
+    num_classes = opts[:num_classes]
+    one_hot = if opts[:one_hot], do: y, else: one_hot_encoding(y, num_classes)
+
+    {_, _, _, _, _, _, final_coeff} =
+      while {iter = 0, x, lr, n, iterations, one_hot,
+             coeff = Nx.broadcast(Nx.tensor(0, type: {:f, 32}), {n, num_classes})},
+            Nx.less(iter, iterations) do
+        coeff = update_coefficients_multinomial(x, one_hot, coeff, lr)
+        {iter + 1, x, lr, n, iterations, one_hot, coeff}
+      end
+
+    %LogisticRegression{coefficients: final_coeff, bias: Nx.tensor(0, type: {:f, 32}), mode: 1}
+  end
+
+  # Normalized softmax
+
+  defnp softmax(t) do
+    normalized = t - Nx.reduce_max(t, axes: [1], keep_axes: true)
+    Nx.transpose(Nx.transpose(Nx.exp(normalized)) / Nx.sum(Nx.exp(normalized), axes: [1]))
+  end
+
+  # Gradient descent for binary regression
+
+  defnp update_coefficients(x, y, {coeff, bias}, lr) do
+    {m, _n} = x.shape
+
+    logit =
+      coeff
+      |> Nx.dot(Nx.transpose(x))
+      |> Nx.add(bias)
+      |> Nx.multiply(-1)
+      |> Nx.exp()
+      |> Nx.add(1)
+      |> then(&(1 / &1))
+
+    diff =
+      logit
+      |> Nx.subtract(Nx.transpose(y))
+      |> Nx.reshape({m})
+
+    coeff_diff =
+      x
+      |> Nx.transpose()
```

Comment:
I think it would be nice to try and avoid these transposes. Perhaps there's a way to get `x` in the desired form beforehand?

Reply:
I decided to precalculate transpositions once and use them later on.